{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Introduction to the `anndata`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## About anndata \n",
    "\n",
    "\n",
    "**What is AnnData and why is it the main data structure of `alphatools`?**\n",
    "\n",
    "[`AnnData`](https://anndata.readthedocs.io/en/stable/) is a Python data for high-dimensional biological data analysis. Originally developed for single-cell genomics, it is increasingly adapted in the scientific ecosystem for the analysis of `omics` data. It forms the central data structure in the [`scverse`](https://scverse.org), an open source software ecosystem for the analysis of omics data. \n",
    "\n",
    "AnnData solves a key proteomics challenge: managing thousands of protein measurements alongside complex sample and protein metadata. Unlike pandas DataFrames where mixing numeric and categorical data complicates matrix operations, AnnData keeps them aligned but separate. It can keep \n",
    "\n",
    "- **Quantitative measurements** (protein abundances, intensities) for thousands of proteins\n",
    "- **Sample metadata** (experimental conditions, patient demographics, batch information)\n",
    "- **Protein annotations** (gene names, functional categories, pathway memberships)\n",
    "- and **Analysis results** (statistical tests, clustering assignments, quality metrics)\n",
    "\n",
    "at a single place which facilitates handling these large amounts of data. Now let's dive into how AnnData is actually organized to understand why it's so effective for proteomics analysis.\n",
    "\n",
    "\n",
    "### Core components of anndata\n",
    "\n",
    "Here is a schematic of an AnnData object, which is created by the `AnnData` class of the `anndata` package [1]:\n",
    "\n",
    "<div align=\"center\">\n",
    "<img src=\"../_static/images/anndata_schema.svg\" width=\"400\" height=\"300\">\n",
    "</div>\n",
    "\n",
    "For most `alphatools` applications, you'll primarily work with three key components:\n",
    "\n",
    "- **`X`**: The Numeric Expression Matrix <br>\n",
    "A numpy array where rows represent **samples** and columns represent **features** (e.g., proteins, precursors, genes)\n",
    "\n",
    "- **`obs`**: Sample Metadata  <br>\n",
    "A DataFrame where rows are **samples** and columns contain metadata properties (e.g., age, disease state, cohort, batch)\n",
    "\n",
    "- **`var`**: Feature Metadata <br>\n",
    "A DataFrame where rows are **features** and columns contain feature properties (e.g., for proteins: Gene names, GO terms, functional annotations)\n",
    "\n",
    "This structure ensures that when you filter samples or features, all associated metadata automatically stays synchronized, preventing common annotation misalignment issues in analysis workflows.\n",
    "\n",
    "**`alphatools` is built around `anndata`** and has a suite of useful filtering and processing functions, making your analyses simpler, more robust and - importantly - compatible with Scanpy/Scverse downstream packages\n",
    "\n",
    "\n",
    "[1] [AnnData Documentation](https://anndata.readthedocs.io/en/stable/)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Basic AnnData syntax\n",
    "\n",
    "We will now explore how users can create and interact with anndata and finally show how `alphatools` enables the search-engine agnostic creation of `anndata` objects from the raw outputs.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import the relevant modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import anndata as ad\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import alphatools as at"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We'll start by manually creating a small synthetic anndata object. Under the hood, `anndata` uses very familiar data structures such as `numpy` arrays and `pandas` Dataframes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>var_names</th>\n",
       "      <th>gene1</th>\n",
       "      <th>gene2</th>\n",
       "      <th>gene3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>obs_names</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cell1</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell2</th>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "var_names  gene1  gene2  gene3\n",
       "obs_names                     \n",
       "cell1          1      0      0\n",
       "cell2          0      2      3"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>age</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>obs_names</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cell1</th>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell2</th>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           age\n",
       "obs_names     \n",
       "cell1       28\n",
       "cell2       29"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>var_names</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>gene1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene3</th>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: [gene1, gene2, gene3]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "numerical_data = np.array([[1, 0, 0], [0, 2, 3]])\n",
    "\n",
    "sample_metadata = pd.DataFrame(\n",
    "    {\n",
    "        \"obs_names\": [\"cell1\", \"cell2\"],\n",
    "        \"age\": [28, 29],\n",
    "    }\n",
    ").set_index(\"obs_names\")\n",
    "\n",
    "feature_metadata = pd.DataFrame({\"var_names\": [\"gene1\", \"gene2\", \"gene3\"]}).set_index(\"var_names\")\n",
    "\n",
    "# Generate AnnData object\n",
    "adata = ad.AnnData(\n",
    "    X=numerical_data,\n",
    "    obs=sample_metadata,\n",
    "    var=feature_metadata,\n",
    ")\n",
    "\n",
    "# We can get a dataframe back\n",
    "df = adata.to_df()\n",
    "display(df)\n",
    "\n",
    "# And also look at the sample and feature metadata\n",
    "display(adata.obs)\n",
    "display(adata.var)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Conversion to other data structures\n",
    "To *get out* of AnnData and back to the more familiar world of dataframes, just run the builtin `.to_df()` function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>var_names</th>\n",
       "      <th>gene1</th>\n",
       "      <th>gene2</th>\n",
       "      <th>gene3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>obs_names</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cell1</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell2</th>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "var_names  gene1  gene2  gene3\n",
       "obs_names                     \n",
       "cell1          1      0      0\n",
       "cell2          0      2      3"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene1</th>\n",
       "      <th>gene2</th>\n",
       "      <th>gene3</th>\n",
       "      <th>age</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>obs_names</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cell1</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell2</th>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene1  gene2  gene3  age\n",
       "obs_names                          \n",
       "cell1          1      0      0   28\n",
       "cell2          0      2      3   29"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df = adata.to_df()\n",
    "display(df)\n",
    "\n",
    "# Get your sample metadata added as columns\n",
    "df = df.join(adata.obs)\n",
    "display(df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## How to load search engine data into AnnData with `alphatools`?\n",
    "\n",
    "The functionality relies on the `alphabase` backbone of PSM and PG readers, which allows for loading and parsing of common search engine output formats. In this notebook we will look at reading data for three common search engines: DIANN, AlphaDIA and Spectronaut. Search engines output data in two main ways: Either as a *long table*, where rows are individual precursors in their respective samples:\n",
    "\n",
    "### PSM tables\n",
    "\n",
    "Peptide spectrum match (PSM) tables are the primary output of proteomics search engines and typically returned in a *long format*. In a PSM table, each row typically represents a single peptide-spectrum-match, i.e. a peptide sequence that the proteomics search engine identified to be compatible with an observed mass spectrum in a given sample. PSM tables contain information about both 1) the peptide sequence, 2) the spectrum, as well as 3) the score assigned to the PSM by the search engine.\n",
    "\n",
    "An example for a PSM table could be something like this\n",
    "\n",
    "\n",
    "| Precursor | Run | Stripped.Sequence | ... |\n",
    "|-----------|-----|-------------------|-----|\n",
    "| PEPTIDEK1 | file_1.raw | PEPTIDEK | ... |\n",
    "| PEPTIDERK2 | file_1.raw | PEPTIDERK | ... |\n",
    "| PEPTIDR2 | file_1.raw | PEPTIDR | ... |\n",
    "| PEPTIDEK1 | file_2.raw | PEPTIDEK | ... |\n",
    "| PEPTIDERK2 | file_2.raw | PEPTIDERK | ... |\n",
    "| PEPTIDR2 | file_2.raw | PEPTIDR | ... |\n",
    "| PEPTIDEK1 | file_3.raw | PEPTIDEK | ... |\n",
    "| PEPTIDERK2 | file_3.raw | PEPTIDERK | ... |\n",
    "| PEPTIDR2 | file_3.raw | PEPTIDR | ... |\n",
    "\n",
    "### Protein group tables\n",
    "\n",
    "Protein group tables are the primary output for protein-level quantification in proteomics workflows. After search engines identify peptide spectrum matches, they aggregate peptide-level evidence to infer protein-level abundances. These protein group tables represent a structured matrix in the *wide format* that maps protein groups (features) to samples (observations), with estimated intensity values as entries.\n",
    "\n",
    "| Protein | file_1.raw | file_2.raw | file_3.raw | ... |\n",
    "|---------|------------|------------|------------|-----|\n",
    "| PROT1 | 1000 | 1200 | 1100 | ... |\n",
    "| PROT2 | 2000 | 2200 | 2100 | ... |\n",
    "| PROT3 | 1500 | 1600 | 1550 | ... |\n",
    "| ... | ... | ... | ... | ... |\n",
    "\n",
    "\n",
    "Typically, downstream analysis tasks are performed on a sample level. Therefore, we are usually interested in a transposed version of the wide format, where rows correspond to samples and columns to features. **Alphatools allows users to easily generate compliant `anndata` objects from raw search engine outputs with its I/O functionalities**."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Getting the example data\n",
    "\n",
    "Small versions of larger real-world datasets are stored right in this repository: `alphatools/docs/_example_data/pg_tables` for wide format protein reports and `alphatools/docs/_example_data/psm_tables` for long format precursor reports. Data for three DIA search engines (DIANN, AlphaDIA and Spectronaut) are deposited there"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Protein Group Tables\n",
    "\n",
    "Let's first take a look a the protein group tables.\n",
    "\n",
    "When you inspect the raw files you will notice that [protein group table outputs vary significantly between search engines](https://alphabase.readthedocs.io/en/latest/tutorials/tutorial_pg_reader.html) with respect to the features and quantities they contain. `alphatools` provides a single function `alphatools.io.read_pg_table()` to read any of the widely used search engine reports into the streamlined `anndata` format.\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You only need to specify the path to the raw output and the type of search engine that generated the output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 61 × 100"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alphadia_pg_anndata = at.io.read_pg_table(\n",
    "    path=\"../_example_data/pg_tables/alphadia/pg.matrix_top100.tsv\",\n",
    "    search_engine=\"alphadia\",\n",
    ")\n",
    "\n",
    "alphadia_pg_anndata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>uniprot_ids</th>\n",
       "      <th>A0A024R6N5;A0A0G2JRN3</th>\n",
       "      <th>A0A075B6H7</th>\n",
       "      <th>A0A075B6H9</th>\n",
       "      <th>A0A075B6I0</th>\n",
       "      <th>A0A075B6I1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleE01</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>2.098375e+07</td>\n",
       "      <td>1.603786e+09</td>\n",
       "      <td>1.234308e+09</td>\n",
       "      <td>9.981811e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleB05</th>\n",
       "      <td>5.248854e+07</td>\n",
       "      <td>1.558676e+07</td>\n",
       "      <td>1.264443e+09</td>\n",
       "      <td>1.017132e+09</td>\n",
       "      <td>1.438419e+08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleB11</th>\n",
       "      <td>3.063311e+08</td>\n",
       "      <td>7.633061e+06</td>\n",
       "      <td>9.859006e+08</td>\n",
       "      <td>6.482135e+08</td>\n",
       "      <td>5.423701e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleA12</th>\n",
       "      <td>1.558016e+08</td>\n",
       "      <td>1.442227e+07</td>\n",
       "      <td>3.558411e+08</td>\n",
       "      <td>1.124133e+09</td>\n",
       "      <td>8.751671e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleB03</th>\n",
       "      <td>2.585067e+08</td>\n",
       "      <td>4.627953e+07</td>\n",
       "      <td>1.224894e+09</td>\n",
       "      <td>8.738715e+08</td>\n",
       "      <td>8.460091e+07</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "uniprot_ids                                         A0A024R6N5;A0A0G2JRN3  \\\n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...           0.000000e+00   \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...           5.248854e+07   \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...           3.063311e+08   \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...           1.558016e+08   \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...           2.585067e+08   \n",
       "\n",
       "uniprot_ids                                           A0A075B6H7  \\\n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...  2.098375e+07   \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...  1.558676e+07   \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...  7.633061e+06   \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...  1.442227e+07   \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...  4.627953e+07   \n",
       "\n",
       "uniprot_ids                                           A0A075B6H9  \\\n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...  1.603786e+09   \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...  1.264443e+09   \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...  9.859006e+08   \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...  3.558411e+08   \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...  1.224894e+09   \n",
       "\n",
       "uniprot_ids                                           A0A075B6I0    A0A075B6I1  \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...  1.234308e+09  9.981811e+07  \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...  1.017132e+09  1.438419e+08  \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...  6.482135e+08  5.423701e+07  \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...  1.124133e+09  8.751671e+07  \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...  8.738715e+08  8.460091e+07  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleE01</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleB05</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleB11</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleA12</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleB03</th>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: [20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleE01, 20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleB05, 20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleB11, 20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleA12, 20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleB03]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>uniprot_ids</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A0A024R6N5;A0A0G2JRN3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A0A075B6H7</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A0A075B6H9</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A0A075B6I0</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A0A075B6I1</th>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: [A0A024R6N5;A0A0G2JRN3, A0A075B6H7, A0A075B6H9, A0A075B6I0, A0A075B6I1]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Inspect the data\n",
    "display(alphadia_pg_anndata.to_df().iloc[:5, :5])\n",
    "display(alphadia_pg_anndata.obs.head())\n",
    "display(alphadia_pg_anndata.var.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DIANN\n",
    "\n",
    "Similarly you can proceed with alternative search engine outputs like DIANN's protein group table. Note that DIANN outputs additional feature-level metadata compared to alphadia that is stored in the `anndata.AnnData.var` attribute."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>proteins</th>\n",
       "      <th>LV469_HUMAN</th>\n",
       "      <th>LV861_HUMAN</th>\n",
       "      <th>LV460_HUMAN</th>\n",
       "      <th>LV548_HUMAN</th>\n",
       "      <th>LV746_HUMAN</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>/fs/gpfs41/lv12/fileset02/pool/pool-mann-projects/alphatools/alzheimers_dataset/RAWDATA/20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleE01.raw</th>\n",
       "      <td>77220500.0</td>\n",
       "      <td>103374000.0</td>\n",
       "      <td>10956700.0</td>\n",
       "      <td></td>\n",
       "      <td>125708000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>/fs/gpfs41/lv12/fileset02/pool/pool-mann-projects/alphatools/alzheimers_dataset/RAWDATA/20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleB05.raw</th>\n",
       "      <td>39566800.0</td>\n",
       "      <td>71580200.0</td>\n",
       "      <td>12655900.0</td>\n",
       "      <td></td>\n",
       "      <td>55909900.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>/fs/gpfs41/lv12/fileset02/pool/pool-mann-projects/alphatools/alzheimers_dataset/RAWDATA/20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleB11.raw</th>\n",
       "      <td>27392100.0</td>\n",
       "      <td>42695000.0</td>\n",
       "      <td>4543130.0</td>\n",
       "      <td></td>\n",
       "      <td>89816500.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>/fs/gpfs41/lv12/fileset02/pool/pool-mann-projects/alphatools/alzheimers_dataset/RAWDATA/20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleA12.raw</th>\n",
       "      <td>12829700.0</td>\n",
       "      <td>55057200.0</td>\n",
       "      <td>7905720.0</td>\n",
       "      <td>202590.0</td>\n",
       "      <td>53537600.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>/fs/gpfs41/lv12/fileset02/pool/pool-mann-projects/alphatools/alzheimers_dataset/RAWDATA/20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleB03.raw</th>\n",
       "      <td>33979400.0</td>\n",
       "      <td>66161200.0</td>\n",
       "      <td>4035670.0</td>\n",
       "      <td></td>\n",
       "      <td>112146000.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "proteins                                           LV469_HUMAN  LV861_HUMAN  \\\n",
       "/fs/gpfs41/lv12/fileset02/pool/pool-mann-projec...  77220500.0  103374000.0   \n",
       "/fs/gpfs41/lv12/fileset02/pool/pool-mann-projec...  39566800.0   71580200.0   \n",
       "/fs/gpfs41/lv12/fileset02/pool/pool-mann-projec...  27392100.0   42695000.0   \n",
       "/fs/gpfs41/lv12/fileset02/pool/pool-mann-projec...  12829700.0   55057200.0   \n",
       "/fs/gpfs41/lv12/fileset02/pool/pool-mann-projec...  33979400.0   66161200.0   \n",
       "\n",
       "proteins                                           LV460_HUMAN LV548_HUMAN  \\\n",
       "/fs/gpfs41/lv12/fileset02/pool/pool-mann-projec...  10956700.0               \n",
       "/fs/gpfs41/lv12/fileset02/pool/pool-mann-projec...  12655900.0               \n",
       "/fs/gpfs41/lv12/fileset02/pool/pool-mann-projec...   4543130.0               \n",
       "/fs/gpfs41/lv12/fileset02/pool/pool-mann-projec...   7905720.0    202590.0   \n",
       "/fs/gpfs41/lv12/fileset02/pool/pool-mann-projec...   4035670.0               \n",
       "\n",
       "proteins                                            LV746_HUMAN  \n",
       "/fs/gpfs41/lv12/fileset02/pool/pool-mann-projec...  125708000.0  \n",
       "/fs/gpfs41/lv12/fileset02/pool/pool-mann-projec...   55909900.0  \n",
       "/fs/gpfs41/lv12/fileset02/pool/pool-mann-projec...   89816500.0  \n",
       "/fs/gpfs41/lv12/fileset02/pool/pool-mann-projec...   53537600.0  \n",
       "/fs/gpfs41/lv12/fileset02/pool/pool-mann-projec...  112146000.0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>/fs/gpfs41/lv12/fileset02/pool/pool-mann-projects/alphatools/alzheimers_dataset/RAWDATA/20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleE01.raw</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>/fs/gpfs41/lv12/fileset02/pool/pool-mann-projects/alphatools/alzheimers_dataset/RAWDATA/20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleB05.raw</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>/fs/gpfs41/lv12/fileset02/pool/pool-mann-projects/alphatools/alzheimers_dataset/RAWDATA/20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleB11.raw</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>/fs/gpfs41/lv12/fileset02/pool/pool-mann-projects/alphatools/alzheimers_dataset/RAWDATA/20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleA12.raw</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>/fs/gpfs41/lv12/fileset02/pool/pool-mann-projects/alphatools/alzheimers_dataset/RAWDATA/20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleB03.raw</th>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: [/fs/gpfs41/lv12/fileset02/pool/pool-mann-projects/alphatools/alzheimers_dataset/RAWDATA/20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleE01.raw, /fs/gpfs41/lv12/fileset02/pool/pool-mann-projects/alphatools/alzheimers_dataset/RAWDATA/20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleB05.raw, /fs/gpfs41/lv12/fileset02/pool/pool-mann-projects/alphatools/alzheimers_dataset/RAWDATA/20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleB11.raw, /fs/gpfs41/lv12/fileset02/pool/pool-mann-projects/alphatools/alzheimers_dataset/RAWDATA/20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleA12.raw, /fs/gpfs41/lv12/fileset02/pool/pool-mann-projects/alphatools/alzheimers_dataset/RAWDATA/20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleB03.raw]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_ids</th>\n",
       "      <th>genes</th>\n",
       "      <th>description</th>\n",
       "      <th>peptide_count</th>\n",
       "      <th>proteotypic_peptide_count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>proteins</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>LV469_HUMAN</th>\n",
       "      <td>A0A075B6H9</td>\n",
       "      <td>IGLV4-69</td>\n",
       "      <td></td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LV861_HUMAN</th>\n",
       "      <td>A0A075B6I0</td>\n",
       "      <td>IGLV8-61</td>\n",
       "      <td></td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LV460_HUMAN</th>\n",
       "      <td>A0A075B6I1</td>\n",
       "      <td>IGLV4-60</td>\n",
       "      <td></td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LV548_HUMAN</th>\n",
       "      <td>A0A075B6I7</td>\n",
       "      <td>IGLV5-48</td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LV746_HUMAN</th>\n",
       "      <td>A0A075B6I9</td>\n",
       "      <td>IGLV7-46</td>\n",
       "      <td></td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            uniprot_ids     genes description  peptide_count  \\\n",
       "proteins                                                       \n",
       "LV469_HUMAN  A0A075B6H9  IGLV4-69                          2   \n",
       "LV861_HUMAN  A0A075B6I0  IGLV8-61                          3   \n",
       "LV460_HUMAN  A0A075B6I1  IGLV4-60                          2   \n",
       "LV548_HUMAN  A0A075B6I7  IGLV5-48                          1   \n",
       "LV746_HUMAN  A0A075B6I9  IGLV7-46                          3   \n",
       "\n",
       "             proteotypic_peptide_count  \n",
       "proteins                                \n",
       "LV469_HUMAN                          2  \n",
       "LV861_HUMAN                          3  \n",
       "LV460_HUMAN                          2  \n",
       "LV548_HUMAN                          1  \n",
       "LV746_HUMAN                          1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Using the pg reader\n",
    "diann_pg_anndata = at.io.read_pg_table(\n",
    "    path=\"../_example_data/pg_tables/diann/pg.matrix_top100.tsv\",\n",
    "    search_engine=\"diann\",\n",
    ")\n",
    "\n",
    "# Inspect the data\n",
    "display(diann_pg_anndata.to_df().iloc[:5, :5])\n",
    "display(diann_pg_anndata.obs.head())\n",
    "display(diann_pg_anndata.var.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Spectronaut"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sn_pg_anndata = at.io.read_pg_table(\n",
    "    path=\"../_example_data/_not_track/spectronaut/pg_table/spectronaut_protein_table.tsv\",\n",
    "    search_engine=\"spectronaut\",\n",
    ")\n",
    "\n",
    "# Inspect the data\n",
    "display(sn_pg_anndata.to_df().iloc[:5, :5])\n",
    "display(sn_pg_anndata.obs.head())\n",
    "display(sn_pg_anndata.var.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Protein data from precursor spectrum match tables\n",
    "\n",
    "You can similarly extract the protein intensity information from PSM tables with the `alphatools.io.read_psm_table` function. Note that the information from the pg tables and psm tables might differ. \n",
    "\n",
    "When might you use PSM tables instead of PG tables? You might choose PSM tables when you need custom protein quantification strategies that differ from the search engine's default approach, when you want to perform quality control at the precursor level to identify problematic peptides, when you're integrating protein data with peptide-level analyses, or when protein group tables aren't available from your search engine.\n",
    "\n",
    "We explore again examples for PSM reports from different search engines."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### AlphaDIA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 60 × 19"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ad_psm_andata = at.io.read_psm_table(\n",
    "    file_paths=\"../_example_data/psm_tables/alphadia/top20_precursors.tsv\",\n",
    "    search_engine=\"alphadia\",\n",
    ")\n",
    "ad_psm_andata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>proteins</th>\n",
       "      <th>P01009;A0A024R6N5</th>\n",
       "      <th>P01011;A0A087WY93;G3V3A0;G3V595</th>\n",
       "      <th>P01024;A0A8Q3SI05;A0A8Q3SI34;A0A8Q3WM02</th>\n",
       "      <th>P02675;D6REL8</th>\n",
       "      <th>P02787</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>raw_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleA01</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5.122159e+09</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.383408e+11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleA02</th>\n",
       "      <td>3.942457e+09</td>\n",
       "      <td>1.895097e+09</td>\n",
       "      <td>7.004725e+09</td>\n",
       "      <td>5.359053e+09</td>\n",
       "      <td>1.166750e+11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleA03</th>\n",
       "      <td>3.670290e+09</td>\n",
       "      <td>1.466623e+09</td>\n",
       "      <td>1.070513e+10</td>\n",
       "      <td>4.963886e+09</td>\n",
       "      <td>1.096607e+11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleA04</th>\n",
       "      <td>5.022584e+09</td>\n",
       "      <td>1.653896e+09</td>\n",
       "      <td>6.425221e+09</td>\n",
       "      <td>5.095784e+09</td>\n",
       "      <td>1.073282e+11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleA05</th>\n",
       "      <td>3.314550e+09</td>\n",
       "      <td>1.044103e+09</td>\n",
       "      <td>5.690721e+09</td>\n",
       "      <td>4.237927e+09</td>\n",
       "      <td>8.503823e+10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "proteins                                            P01009;A0A024R6N5  \\\n",
       "raw_name                                                                \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...                NaN   \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...       3.942457e+09   \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...       3.670290e+09   \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...       5.022584e+09   \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...       3.314550e+09   \n",
       "\n",
       "proteins                                            P01011;A0A087WY93;G3V3A0;G3V595  \\\n",
       "raw_name                                                                              \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...                              NaN   \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...                     1.895097e+09   \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...                     1.466623e+09   \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...                     1.653896e+09   \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...                     1.044103e+09   \n",
       "\n",
       "proteins                                            P01024;A0A8Q3SI05;A0A8Q3SI34;A0A8Q3WM02  \\\n",
       "raw_name                                                                                      \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...                             5.122159e+09   \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...                             7.004725e+09   \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...                             1.070513e+10   \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...                             6.425221e+09   \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...                             5.690721e+09   \n",
       "\n",
       "proteins                                            P02675;D6REL8  \\\n",
       "raw_name                                                            \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...            NaN   \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...   5.359053e+09   \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...   4.963886e+09   \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...   5.095784e+09   \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...   4.237927e+09   \n",
       "\n",
       "proteins                                                  P02787  \n",
       "raw_name                                                          \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...  1.383408e+11  \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...  1.166750e+11  \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...  1.096607e+11  \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...  1.073282e+11  \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...  8.503823e+10  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>raw_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleA01</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleA02</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleA03</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleA04</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleA05</th>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: [20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleA01, 20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleA02, 20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleA03, 20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleA04, 20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleA05]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>proteins</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>P01009;A0A024R6N5</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P01011;A0A087WY93;G3V3A0;G3V595</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P01024;A0A8Q3SI05;A0A8Q3SI34;A0A8Q3WM02</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P02675;D6REL8</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P02787</th>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: [P01009;A0A024R6N5, P01011;A0A087WY93;G3V3A0;G3V595, P01024;A0A8Q3SI05;A0A8Q3SI34;A0A8Q3WM02, P02675;D6REL8, P02787]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Inspect the data\n",
    "display(ad_psm_andata.to_df().iloc[:5, :5])\n",
    "display(ad_psm_andata.obs.head())\n",
    "display(ad_psm_andata.var.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DIANN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>proteins</th>\n",
       "      <th>A4-11_HUMAN;A4-8_HUMAN;A4_HUMAN</th>\n",
       "      <th>AACT_HUMAN</th>\n",
       "      <th>ALBU_HUMAN</th>\n",
       "      <th>APOE_HUMAN</th>\n",
       "      <th>CLUS-6_HUMAN</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>raw_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleA01</th>\n",
       "      <td>8.406033e+08</td>\n",
       "      <td>9.798148e+08</td>\n",
       "      <td>1.929707e+10</td>\n",
       "      <td>6.036072e+09</td>\n",
       "      <td>7.492970e+09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleA02</th>\n",
       "      <td>8.213831e+08</td>\n",
       "      <td>1.396344e+09</td>\n",
       "      <td>2.287499e+10</td>\n",
       "      <td>7.632371e+09</td>\n",
       "      <td>7.724988e+09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleA03</th>\n",
       "      <td>8.760188e+08</td>\n",
       "      <td>1.025414e+09</td>\n",
       "      <td>1.264076e+10</td>\n",
       "      <td>4.464353e+09</td>\n",
       "      <td>6.336454e+09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleA04</th>\n",
       "      <td>1.073826e+09</td>\n",
       "      <td>1.272798e+09</td>\n",
       "      <td>1.557733e+10</td>\n",
       "      <td>6.688002e+09</td>\n",
       "      <td>6.312889e+09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleA05</th>\n",
       "      <td>1.120733e+09</td>\n",
       "      <td>8.905130e+08</td>\n",
       "      <td>1.579670e+10</td>\n",
       "      <td>4.973349e+09</td>\n",
       "      <td>5.925785e+09</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "proteins                                            A4-11_HUMAN;A4-8_HUMAN;A4_HUMAN  \\\n",
       "raw_name                                                                              \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...                     8.406033e+08   \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...                     8.213831e+08   \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...                     8.760188e+08   \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...                     1.073826e+09   \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...                     1.120733e+09   \n",
       "\n",
       "proteins                                              AACT_HUMAN  \\\n",
       "raw_name                                                           \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...  9.798148e+08   \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...  1.396344e+09   \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...  1.025414e+09   \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...  1.272798e+09   \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...  8.905130e+08   \n",
       "\n",
       "proteins                                              ALBU_HUMAN  \\\n",
       "raw_name                                                           \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...  1.929707e+10   \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...  2.287499e+10   \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...  1.264076e+10   \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...  1.557733e+10   \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...  1.579670e+10   \n",
       "\n",
       "proteins                                              APOE_HUMAN  CLUS-6_HUMAN  \n",
       "raw_name                                                                        \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...  6.036072e+09  7.492970e+09  \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...  7.632371e+09  7.724988e+09  \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...  4.464353e+09  6.336454e+09  \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...  6.688002e+09  6.312889e+09  \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...  4.973349e+09  5.925785e+09  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>raw_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleA01</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleA02</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleA03</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleA04</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleA05</th>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: [20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleA01, 20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleA02, 20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleA03, 20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleA04, 20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleA05]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>proteins</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A4-11_HUMAN;A4-8_HUMAN;A4_HUMAN</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AACT_HUMAN</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ALBU_HUMAN</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>APOE_HUMAN</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CLUS-6_HUMAN</th>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: [A4-11_HUMAN;A4-8_HUMAN;A4_HUMAN, AACT_HUMAN, ALBU_HUMAN, APOE_HUMAN, CLUS-6_HUMAN]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "diann_pg_andata = at.io.read_psm_table(\n",
    "    file_paths=\"../_example_data/psm_tables/diann/top20_report.parquet\",\n",
    "    search_engine=\"diann\",\n",
    ")\n",
    "\n",
    "# Inspect the data\n",
    "display(diann_pg_andata.to_df().iloc[:5, :5])\n",
    "display(diann_pg_andata.obs.head())\n",
    "display(diann_pg_andata.var.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Spectronaut"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# sn_pg_andata = at.io.read_psm_table(\n",
    "#     file_paths=\"../_example_data/psm_tables/spectronaut/example_dataset_mouse_sn_top20peptides.tsv\",\n",
    "#     search_engine=\"spectronaut\",\n",
    "#     intensity_column = \"F.PeakArea\",\n",
    "#     feature_id_column = \"FG.Id\"\n",
    "# )\n",
    "\n",
    "# # Inspect the data\n",
    "# display(sn_pg_andata.to_df().iloc[:5, :5])\n",
    "# display(sn_pg_andata.obs.head())\n",
    "# display(sn_pg_andata.var.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## In Summary, we learned...\n",
    "\n",
    "- How AnnData can help us keep our data and metadata aligned\n",
    "- How to generate AnnData objects from our input dataframes\n",
    "- How to get dataframes back from AnnData objects\n",
    "- How to load protein tables using `alphatools` readers\n",
    "- How to load and pivot PSM tables using `alphatools` readers\n",
    "\n",
    "From here on, proteomics data can be used for all further downstream analyses. Continue with notebook [`/docs/notebooks/01_basic_workflow.ipynb`] to learn how to merge your metadata into AnnData objects and perform filtering and exploratory data analysis!"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "alphatools-dev",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
