{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# What is AnnData and why is it the main data structure of `alphatools`?\n",
    "\n",
    "AnnData is a Python data format from the Scverse (open source single-cell software ecosystem) that keeps numeric data (arrays) and metadata neatly aligned together. This solves a core limitation of Pandas DataFrames where numerical and non-numerical data must coexist in the same structure, making dataframe-wide numeric operations difficult. below is a schematic of an AnnData object, which is created by the `AnnData` class of the `anndata` package [1]:\n",
    "\n",
    "<div align=\"center\">\n",
    "<img src=\"../../assets/anndata_schema.svg\" width=\"400\" height=\"300\">\n",
    "</div>\n",
    "\n",
    "## Core Components for AlphaTools\n",
    "\n",
    "For most `alphatools` applications, you'll primarily work with three key components:\n",
    "\n",
    "### **X**: The Numeric Expression Matrix\n",
    "A numpy array where rows represent **samples** and columns represent **features** (e.g., proteins, precursors, genes)\n",
    "\n",
    "### **obs**: Sample Metadata  \n",
    "A DataFrame where rows are **samples** and columns contain metadata properties (e.g., age, disease state, cohort, batch)\n",
    "\n",
    "### **var**: Feature Metadata\n",
    "A DataFrame where rows are **features** and columns contain feature properties (e.g., for proteins: Gene names, GO terms, functional annotations)\n",
    "\n",
    "This structure ensures that when you filter samples or features, all associated metadata automatically stays synchronized, preventing common annotation misalignment issues in analysis workflows.\n",
    "\n",
    "## References\n",
    "\n",
    "- [1] [AnnData Documentation](https://anndata.readthedocs.io/en/stable/)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "UsageError: Missing module name.\n"
     ]
    }
   ],
   "source": [
    "%load_ext\n",
    "%autoreload 2\n",
    "\n",
    "import anndata as ad\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "numerical_data = np.array([[1, 0, 0], [0, 2, 3]])\n",
    "\n",
    "sample_metadata = pd.DataFrame({\"obs_names\": [\"cell1\", \"cell2\"]}).set_index(\"obs_names\")\n",
    "\n",
    "feature_metadata = pd.DataFrame({\"var_names\": [\"gene1\", \"gene2\", \"gene3\"]}).set_index(\"var_names\")\n",
    "\n",
    "# Generate AnnData object\n",
    "adata = ad.AnnData(\n",
    "    X=numerical_data,\n",
    "    obs=sample_metadata,\n",
    "    var=feature_metadata,\n",
    ")\n",
    "\n",
    "# We can get a dataframe back\n",
    "df = adata.to_df()\n",
    "display(df)\n",
    "\n",
    "# And also look at the sample and feature metadata\n",
    "display(adata.obs)\n",
    "display(adata.var)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### `alphatools` has a suite of useful filtering and processing functions built around AnnData, making your analyses simpler, more robust and - importantly - compatible with Scanpy/Scverse functions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## How do I load search engine data into AnnData with `alphatools`?\n",
    "\n",
    "This notebook demonstrates core `alphatools` functionality to load protein and precursor data from different sources. \n",
    "\n",
    "The functionality relies on the `alphabase` backbone of PSM and PG readers, which allows for loading and parsing of common search engine output formats. In this notebook we will look at reading data for three common search engines: DIANN, AlphaDIA and Spectronaut.\n",
    "\n",
    "### What kind of data do we expect?\n",
    "\n",
    "Search engines output data in two main ways: Either as a *long table*, where rows are individual precursors in their respective samples:\n",
    "\n",
    "### *PSM report* (long format)\n",
    "\n",
    "| Precursor | Run | Stripped.Sequence | ... |\n",
    "|-----------|-----|-------------------|-----|\n",
    "| PEPTIDEK1 | file_1.raw | PEPTIDEK | ... |\n",
    "| PEPTIDERK2 | file_1.raw | PEPTIDERK | ... |\n",
    "| PEPTIDR2 | file_1.raw | PEPTIDR | ... |\n",
    "| PEPTIDEK1 | file_2.raw | PEPTIDEK | ... |\n",
    "| PEPTIDERK2 | file_2.raw | PEPTIDERK | ... |\n",
    "| PEPTIDR2 | file_2.raw | PEPTIDR | ... |\n",
    "| PEPTIDEK1 | file_3.raw | PEPTIDEK | ... |\n",
    "| PEPTIDERK2 | file_3.raw | PEPTIDERK | ... |\n",
    "| PEPTIDR2 | file_3.raw | PEPTIDR | ... |\n",
    "\n",
    "### *Protein group report* (wide format)\n",
    "\n",
    "| Protein | file_1.raw | file_2.raw | file_3.raw | ... |\n",
    "|---------|------------|------------|------------|-----|\n",
    "| PROT1 | 1000 | 1200 | 1100 | ... |\n",
    "| PROT2 | 2000 | 2200 | 2100 | ... |\n",
    "| PROT3 | 1500 | 1600 | 1550 | ... |\n",
    "| ... | ... | ... | ... | ... |\n",
    "\n",
    "--> 99 % of the time for data analysis, we are interested in a (transposed) version of the wide format, but sometimes we need to transform the long table into a wide format. Both can be done with `alphatools` and the underlying `alphabase`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Getting the example data\n",
    "\n",
    "Small versions of larger real-world datasets are stored right in this repository: `alphatools/docs/_example_data/pg_tables` for wide format protein reports and `alphatools/docs/_example_data/psm_tables` for long format precursor reports. Data for three DIA search engines (DIANN, AlphaDIA and Spectronaut) are deposited there\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:rdkit:Enabling RDKit 2024.03.3 jupyter extensions\n"
     ]
    }
   ],
   "source": [
    "from alphatools.io import pg_reader"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Wide format Protein Tables"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### AlphaDIA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>uniprot_ids</th>\n",
       "      <th>A0A024R6N5;A0A0G2JRN3</th>\n",
       "      <th>A0A075B6H7</th>\n",
       "      <th>A0A075B6H9</th>\n",
       "      <th>A0A075B6I0</th>\n",
       "      <th>A0A075B6I1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleE01</th>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>2.098375e+07</td>\n",
       "      <td>1.603786e+09</td>\n",
       "      <td>1.234308e+09</td>\n",
       "      <td>9.981811e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleB05</th>\n",
       "      <td>5.248854e+07</td>\n",
       "      <td>1.558676e+07</td>\n",
       "      <td>1.264443e+09</td>\n",
       "      <td>1.017132e+09</td>\n",
       "      <td>1.438419e+08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleB11</th>\n",
       "      <td>3.063311e+08</td>\n",
       "      <td>7.633061e+06</td>\n",
       "      <td>9.859006e+08</td>\n",
       "      <td>6.482135e+08</td>\n",
       "      <td>5.423701e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleA12</th>\n",
       "      <td>1.558016e+08</td>\n",
       "      <td>1.442227e+07</td>\n",
       "      <td>3.558411e+08</td>\n",
       "      <td>1.124133e+09</td>\n",
       "      <td>8.751671e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1_sampleB03</th>\n",
       "      <td>2.585067e+08</td>\n",
       "      <td>4.627953e+07</td>\n",
       "      <td>1.224894e+09</td>\n",
       "      <td>8.738715e+08</td>\n",
       "      <td>8.460091e+07</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "uniprot_ids                                         A0A024R6N5;A0A0G2JRN3  \\\n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...           0.000000e+00   \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...           5.248854e+07   \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...           3.063311e+08   \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...           1.558016e+08   \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...           2.585067e+08   \n",
       "\n",
       "uniprot_ids                                           A0A075B6H7  \\\n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...  2.098375e+07   \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...  1.558676e+07   \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...  7.633061e+06   \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...  1.442227e+07   \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...  4.627953e+07   \n",
       "\n",
       "uniprot_ids                                           A0A075B6H9  \\\n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...  1.603786e+09   \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...  1.264443e+09   \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...  9.859006e+08   \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...  3.558411e+08   \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...  1.224894e+09   \n",
       "\n",
       "uniprot_ids                                           A0A075B6I0    A0A075B6I1  \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...  1.234308e+09  9.981811e+07  \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...  1.017132e+09  1.438419e+08  \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...  6.482135e+08  5.423701e+07  \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...  1.124133e+09  8.751671e+07  \n",
       "20180618_QX0_JaBa_SA_LC12_5_CSF1_1_8-1xD1xS1fM1...  8.738715e+08  8.460091e+07  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "# Using the pg reader\n",
    "alphadia_pg_anndata = pg_reader.read_pg_table(\n",
    "    path=\"../_example_data/pg_tables/alphadia/pg.matrix_top100.tsv\",\n",
    "    search_engine=\"alphadia\",\n",
    ")\n",
    "\n",
    "# Inspect the data\n",
    "display(alphadia_pg_anndata.to_df().iloc[:5, :5])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DIANN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/vincenthbrennsteiner/miniconda3/envs/alphaverse/lib/python3.11/site-packages/anndata/_core/aligned_df.py:68: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n"
     ]
    },
    {
     "ename": "TypeError",
     "evalue": "Setting a MultiIndex dtype to anything other than object is not supported",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m/Users/vincenthbrennsteiner/Documents/mann_labs/_git_repositories/alphatools/docs/notebooks/00_data_sources.ipynb Cell 11\u001b[0m line \u001b[0;36m2\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/vincenthbrennsteiner/Documents/mann_labs/_git_repositories/alphatools/docs/notebooks/00_data_sources.ipynb#Y135sZmlsZQ%3D%3D?line=0'>1</a>\u001b[0m \u001b[39m# Using the pg reader\u001b[39;00m\n\u001b[0;32m----> <a href='vscode-notebook-cell:/Users/vincenthbrennsteiner/Documents/mann_labs/_git_repositories/alphatools/docs/notebooks/00_data_sources.ipynb#Y135sZmlsZQ%3D%3D?line=1'>2</a>\u001b[0m pg_reader\u001b[39m.\u001b[39;49mread_pg_table(\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/vincenthbrennsteiner/Documents/mann_labs/_git_repositories/alphatools/docs/notebooks/00_data_sources.ipynb#Y135sZmlsZQ%3D%3D?line=2'>3</a>\u001b[0m     path\u001b[39m=\u001b[39;49m\u001b[39m\"\u001b[39;49m\u001b[39m../_example_data/pg_tables/diann/pg.matrix_top100.tsv\u001b[39;49m\u001b[39m\"\u001b[39;49m,\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/vincenthbrennsteiner/Documents/mann_labs/_git_repositories/alphatools/docs/notebooks/00_data_sources.ipynb#Y135sZmlsZQ%3D%3D?line=3'>4</a>\u001b[0m     search_engine\u001b[39m=\u001b[39;49m\u001b[39m\"\u001b[39;49m\u001b[39mdiann\u001b[39;49m\u001b[39m\"\u001b[39;49m,\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/vincenthbrennsteiner/Documents/mann_labs/_git_repositories/alphatools/docs/notebooks/00_data_sources.ipynb#Y135sZmlsZQ%3D%3D?line=4'>5</a>\u001b[0m )\n",
      "File \u001b[0;32m~/Documents/mann_labs/_git_repositories/alphatools/src/alphatools/io/pg_reader.py:114\u001b[0m, in \u001b[0;36mread_pg_table\u001b[0;34m(path, search_engine, column_mapping, measurement_regex, **reader_provider_kwargs)\u001b[0m\n\u001b[1;32m    111\u001b[0m     var_df \u001b[39m=\u001b[39m df\u001b[39m.\u001b[39mindex\u001b[39m.\u001b[39mto_frame(index\u001b[39m=\u001b[39m\u001b[39mFalse\u001b[39;00m)\n\u001b[1;32m    113\u001b[0m \u001b[39m# Observations x Features\u001b[39;00m\n\u001b[0;32m--> 114\u001b[0m \u001b[39mreturn\u001b[39;00m ad\u001b[39m.\u001b[39;49mAnnData(X\u001b[39m=\u001b[39;49mdf\u001b[39m.\u001b[39;49mto_numpy()\u001b[39m.\u001b[39;49mT, var\u001b[39m=\u001b[39;49mvar_df, obs\u001b[39m=\u001b[39;49mpd\u001b[39m.\u001b[39;49mDataFrame(index\u001b[39m=\u001b[39;49mdf\u001b[39m.\u001b[39;49mcolumns))\n",
      "File \u001b[0;32m~/miniconda3/envs/alphaverse/lib/python3.11/site-packages/anndata/_core/anndata.py:252\u001b[0m, in \u001b[0;36mAnnData.__init__\u001b[0;34m(self, X, obs, var, uns, obsm, varm, layers, raw, dtype, shape, filename, filemode, asview, obsp, varp, oidx, vidx)\u001b[0m\n\u001b[1;32m    250\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_init_as_view(X, oidx, vidx)\n\u001b[1;32m    251\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[0;32m--> 252\u001b[0m     \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_init_as_actual(\n\u001b[1;32m    253\u001b[0m         X\u001b[39m=\u001b[39;49mX,\n\u001b[1;32m    254\u001b[0m         obs\u001b[39m=\u001b[39;49mobs,\n\u001b[1;32m    255\u001b[0m         var\u001b[39m=\u001b[39;49mvar,\n\u001b[1;32m    256\u001b[0m         uns\u001b[39m=\u001b[39;49muns,\n\u001b[1;32m    257\u001b[0m         obsm\u001b[39m=\u001b[39;49mobsm,\n\u001b[1;32m    258\u001b[0m         varm\u001b[39m=\u001b[39;49mvarm,\n\u001b[1;32m    259\u001b[0m         raw\u001b[39m=\u001b[39;49mraw,\n\u001b[1;32m    260\u001b[0m         layers\u001b[39m=\u001b[39;49mlayers,\n\u001b[1;32m    261\u001b[0m         dtype\u001b[39m=\u001b[39;49mdtype,\n\u001b[1;32m    262\u001b[0m         shape\u001b[39m=\u001b[39;49mshape,\n\u001b[1;32m    263\u001b[0m         obsp\u001b[39m=\u001b[39;49mobsp,\n\u001b[1;32m    264\u001b[0m         varp\u001b[39m=\u001b[39;49mvarp,\n\u001b[1;32m    265\u001b[0m         filename\u001b[39m=\u001b[39;49mfilename,\n\u001b[1;32m    266\u001b[0m         filemode\u001b[39m=\u001b[39;49mfilemode,\n\u001b[1;32m    267\u001b[0m     )\n",
      "File \u001b[0;32m~/miniconda3/envs/alphaverse/lib/python3.11/site-packages/anndata/_core/anndata.py:427\u001b[0m, in \u001b[0;36mAnnData._init_as_actual\u001b[0;34m(self, X, obs, var, uns, obsm, varm, varp, obsp, raw, layers, dtype, shape, filename, filemode)\u001b[0m\n\u001b[1;32m    423\u001b[0m \u001b[39m# annotations\u001b[39;00m\n\u001b[1;32m    424\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_obs \u001b[39m=\u001b[39m _gen_dataframe(\n\u001b[1;32m    425\u001b[0m     obs, [\u001b[39m\"\u001b[39m\u001b[39mobs_names\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mrow_names\u001b[39m\u001b[39m\"\u001b[39m], source\u001b[39m=\u001b[39msource, attr\u001b[39m=\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mobs\u001b[39m\u001b[39m\"\u001b[39m, length\u001b[39m=\u001b[39mn_obs\n\u001b[1;32m    426\u001b[0m )\n\u001b[0;32m--> 427\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_var \u001b[39m=\u001b[39m _gen_dataframe(\n\u001b[1;32m    428\u001b[0m     var, [\u001b[39m\"\u001b[39;49m\u001b[39mvar_names\u001b[39;49m\u001b[39m\"\u001b[39;49m, \u001b[39m\"\u001b[39;49m\u001b[39mcol_names\u001b[39;49m\u001b[39m\"\u001b[39;49m], source\u001b[39m=\u001b[39;49msource, attr\u001b[39m=\u001b[39;49m\u001b[39m\"\u001b[39;49m\u001b[39mvar\u001b[39;49m\u001b[39m\"\u001b[39;49m, length\u001b[39m=\u001b[39;49mn_vars\n\u001b[1;32m    429\u001b[0m )\n\u001b[1;32m    431\u001b[0m \u001b[39m# now we can verify if indices match!\u001b[39;00m\n\u001b[1;32m    432\u001b[0m \u001b[39mfor\u001b[39;00m attr_name, x_name, idx \u001b[39min\u001b[39;00m x_indices:\n",
      "File \u001b[0;32m~/miniconda3/envs/alphaverse/lib/python3.11/functools.py:909\u001b[0m, in \u001b[0;36msingledispatch.<locals>.wrapper\u001b[0;34m(*args, **kw)\u001b[0m\n\u001b[1;32m    905\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m args:\n\u001b[1;32m    906\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mTypeError\u001b[39;00m(\u001b[39mf\u001b[39m\u001b[39m'\u001b[39m\u001b[39m{\u001b[39;00mfuncname\u001b[39m}\u001b[39;00m\u001b[39m requires at least \u001b[39m\u001b[39m'\u001b[39m\n\u001b[1;32m    907\u001b[0m                     \u001b[39m'\u001b[39m\u001b[39m1 positional argument\u001b[39m\u001b[39m'\u001b[39m)\n\u001b[0;32m--> 909\u001b[0m \u001b[39mreturn\u001b[39;00m dispatch(args[\u001b[39m0\u001b[39;49m]\u001b[39m.\u001b[39;49m\u001b[39m__class__\u001b[39;49m)(\u001b[39m*\u001b[39;49margs, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkw)\n",
      "File \u001b[0;32m~/miniconda3/envs/alphaverse/lib/python3.11/site-packages/anndata/_core/aligned_df.py:69\u001b[0m, in \u001b[0;36m_gen_dataframe_df\u001b[0;34m(anno, index_names, source, attr, length)\u001b[0m\n\u001b[1;32m     67\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m is_string_dtype(anno\u001b[39m.\u001b[39mindex):\n\u001b[1;32m     68\u001b[0m     warnings\u001b[39m.\u001b[39mwarn(\u001b[39m\"\u001b[39m\u001b[39mTransforming to str index.\u001b[39m\u001b[39m\"\u001b[39m, ImplicitModificationWarning)\n\u001b[0;32m---> 69\u001b[0m     anno\u001b[39m.\u001b[39mindex \u001b[39m=\u001b[39m anno\u001b[39m.\u001b[39;49mindex\u001b[39m.\u001b[39;49mastype(\u001b[39mstr\u001b[39;49m)\n\u001b[1;32m     70\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mlen\u001b[39m(anno\u001b[39m.\u001b[39mcolumns):\n\u001b[1;32m     71\u001b[0m     anno\u001b[39m.\u001b[39mcolumns \u001b[39m=\u001b[39m anno\u001b[39m.\u001b[39mcolumns\u001b[39m.\u001b[39mastype(\u001b[39mstr\u001b[39m)\n",
      "File \u001b[0;32m~/miniconda3/envs/alphaverse/lib/python3.11/site-packages/pandas/core/indexes/multi.py:3873\u001b[0m, in \u001b[0;36mMultiIndex.astype\u001b[0;34m(self, dtype, copy)\u001b[0m\n\u001b[1;32m   3871\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mNotImplementedError\u001b[39;00m(msg)\n\u001b[1;32m   3872\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m is_object_dtype(dtype):\n\u001b[0;32m-> 3873\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mTypeError\u001b[39;00m(\n\u001b[1;32m   3874\u001b[0m         \u001b[39m\"\u001b[39m\u001b[39mSetting a MultiIndex dtype to anything other than object \u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m   3875\u001b[0m         \u001b[39m\"\u001b[39m\u001b[39mis not supported\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m   3876\u001b[0m     )\n\u001b[1;32m   3877\u001b[0m \u001b[39mif\u001b[39;00m copy \u001b[39mis\u001b[39;00m \u001b[39mTrue\u001b[39;00m:\n\u001b[1;32m   3878\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_view()\n",
      "\u001b[0;31mTypeError\u001b[0m: Setting a MultiIndex dtype to anything other than object is not supported"
     ]
    }
   ],
   "source": [
    "# Using the pg reader\n",
    "pg_reader.read_pg_table(\n",
    "    path=\"../_example_data/pg_tables/diann/pg.matrix_top100.tsv\",\n",
    "    search_engine=\"diann\",\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Spectronaut"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Using the pg reader\n",
    "pg_reader.read_pg_table(\n",
    "    path=None,\n",
    "    search_engine=None,\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "alphaverse",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
