{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Basic `alphatools` workflow\n",
    "\n",
    "1. load data and meta data\n",
    "2. basic minimal preprocessing\n",
    "3. perform PCA\n",
    "4. plot PCA embeddings, variance and loadings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import logging\n",
    "from anndata import AnnData\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "import alphatools as at\n",
    "from alphatools.pl.figure import create_figure, label_axes\n",
    "from alphatools.pl.plots import Plots\n",
    "from alphatools.pp.embeddings import pca\n",
    "from alphatools.pp.metadata import add_core_proteome_mask\n",
    "\n",
    "\n",
    "logging.basicConfig(level=logging.INFO)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load the data \n",
    "\n",
    "Dataset taken from datashare folder: https://datashare.biochem.mpg.de/s/sSYkOj22kM5AJ4O"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Change after merging, and use Lucas' code to load the DIA-NN data\n",
    "\n",
    "# load protein groups into an AnnData object with index & columns as obs & var\n",
    "url_data = \"https://datashare.biochem.mpg.de/public.php/dav/files/NGxD86CQSbRWmzw/report.pg_matrix.tsv\"\n",
    "df = pd.read_csv(url_data, sep=\"\\t\")\n",
    "\n",
    "url_metadata = \"https://datashare.biochem.mpg.de/public.php/dav/files/sSYkOj22kM5AJ4O/simple_metadata.csv\"\n",
    "md = pd.read_csv(url_metadata, sep=\",\")\n",
    "\n",
    "adata = AnnData(\n",
    "    X=df.drop(columns=[\"Protein.Group\", \"Protein.Names\", \"Genes\", \"First.Protein.Description\"]).to_numpy().T,\n",
    "    obs=md,\n",
    "    var=df[[\"Protein.Group\", \"Protein.Names\", \"Genes\", \"First.Protein.Description\"]].set_index(\"Protein.Group\"),\n",
    ")\n",
    "\n",
    "# overview of the data object\n",
    "print(adata)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Basic EDA on a synthetic example dataset:\n",
    "\n",
    "1. Generate example data\n",
    "2. Filter for data completeness on sample level\n",
    "3. Visualize samples as histograms\n",
    "4. Save data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Filter by data completeness:\n",
    "\n",
    "Remove features which have more than the allowed fraction of missing values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"The numeric data in the anndata object:\")\n",
    "display(adata.to_df().head())\n",
    "\n",
    "print(\"The sample-level metadata in the anndata object:\")\n",
    "display(adata.obs.head())\n",
    "\n",
    "print(\"The feature-level metadata in the anndata object:\")\n",
    "display(adata.var.head())\n",
    "\n",
    "#  filter out features with more than 25 % missing values\n",
    "print(\"Before filtering, the shape of the anndata object: \", adata.shape)\n",
    "adata = at.pp.filter_data_completeness(adata=adata, max_missing=0.25)\n",
    "print(\"After filtering, the shape of the anndata object: \", adata.shape)\n",
    "\n",
    "print(\"The numeric data in the anndata object:\")\n",
    "display(adata.to_df().head())\n",
    "\n",
    "print(\"The sample-level metadata in the anndata object:\")\n",
    "display(adata.obs.head())\n",
    "\n",
    "print(\"The feature-level metadata in the anndata object:\")\n",
    "display(adata.var.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Creating new layers prior to preprocessing\n",
    "\n",
    "This way, we can save the raw data and try different pp steps on the raw data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save the raw data before log transformation\n",
    "adata.layers[\"raw\"] = adata.X.copy()\n",
    "\n",
    "# log2 transform the data\n",
    "adata.X = np.log2(adata.X + 1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Visualize the distribution of values in different levels of an observational metadata variable\n",
    "\n",
    "In this example, check the distribution of \"gene_1\" expression values per cell type."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Apply the AxisManager to make axes iterable and apply consistent AlphaTools styling.\n",
    "# Axes can also be accessed directly by indexing the axm object.\n",
    "fig, axm = create_figure(nrows=1, ncols=2, figsize=(5, 3))\n",
    "\n",
    "# Plot.histogram handles adata natively. Columns from the data and metadata are accessible\n",
    "# Focus on the distribution of protein A1L0T0\n",
    "ax = axm.next()\n",
    "Plots.histogram(\n",
    "    data=adata,\n",
    "    value_column=\"A1L0T0\",\n",
    "    bins=20,\n",
    "    legend=\"auto\",\n",
    "    ax=ax,\n",
    "    hist_kwargs={\"alpha\": 0.5, \"histtype\": \"stepfilled\", \"linewidth\": 0.5, \"edgecolor\": \"black\"},\n",
    ")\n",
    "label_axes(ax, \"A1L0T0\", \"Frequency\", \"Distribution of A1L0T0\")\n",
    "\n",
    "# Focus on the distribution of protein A1L0T0 in the different replicates\n",
    "ax = axm.next()\n",
    "Plots.histogram(\n",
    "    data=adata,\n",
    "    value_column=\"A1L0T0\",\n",
    "    color_map_column=\"replicate\",\n",
    "    bins=20,\n",
    "    legend=\"auto\",\n",
    "    ax=ax,\n",
    "    hist_kwargs={\"alpha\": 0.5, \"histtype\": \"stepfilled\", \"linewidth\": 0.5, \"edgecolor\": \"black\"},\n",
    ")\n",
    "label_axes(ax, \"A1L0T0\", \"Frequency\", \"Distribution of A1L0T0 by replicate\")\n",
    "\n",
    "# # save figure\n",
    "# save_figure(\n",
    "#     fig=fig,\n",
    "#     filename=\"sample_histogram.png\",\n",
    "#     output_dir=output_directory,\n",
    "#     dpi=300,\n",
    "#     transparent=False,\n",
    "# )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Running PCA\n",
    "\n",
    "Before running PCA, we need to filter out NaN values. PCA can not be computed on matrices with missing values. Therefore, prior to PCA, we will create a list of 'core proteins' of proteins detected in all observations, save it in the feature meta data frame (adata.var)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add a new column to the adata.var object with the name \"isCore\" to indicate whether the feature is part of the core proteome\n",
    "add_core_proteome_mask(adata, layer=\"raw\", new_column_name=\"is_core\")\n",
    "\n",
    "# view hoe many features are part of the core proteome\n",
    "print(\"The number of features in the core proteome:\")\n",
    "print(adata.var[\"is_core\"].value_counts())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we can run PCA, specifying the adata.var column that filters the proteins by 100% completeness:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# this function is now implemented on sample level (PCA of the observations).\n",
    "pca(adata, meta_data_mask_column_name=\"is_core\", n_comps=10)\n",
    "\n",
    "# view the PCA results\n",
    "print(\"The dimensions of PC coordinates in the adata.obsm are (n_obs x n_comp):\")\n",
    "print(adata.obsm[\"X_pca_obs\"].shape)\n",
    "print(\"The PCA loadings in the adata.varm are (n_var x n_comp):\")\n",
    "print(adata.varm[\"PCs_obs\"].shape)\n",
    "print(\"Ratio of explained variance (n_comp):\")\n",
    "print(adata.uns[\"variance_pca_obs\"][\"variance_ratio\"])\n",
    "print(\"The explained variance (n_comp):\")\n",
    "print(adata.uns[\"variance_pca_obs\"][\"variance\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In addition to running PCA to get a dimentional reduction of the observations (samples), we can also perform PCA on the features (proteins). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now run PCA on the protein space to get their projection in the PCA space.\n",
    "pca(adata, meta_data_mask_column_name=\"is_core\", n_comps=10, dim_space=\"var\")\n",
    "\n",
    "# view the PCA results for features\n",
    "print(\"----- PCA ON FEATURES -----\")\n",
    "print(\"The dimensions of PC coordinates in the adata.varm are (n_obs x n_comp):\")\n",
    "print(adata.varm[\"X_pca_var\"].shape)\n",
    "print(\"The PCA loadings of the samples in the adata.obsm are (n_var x n_comp):\")\n",
    "print(adata.obsm[\"PCs_var\"].shape)\n",
    "print(\"Ratio of explained variance (n_comp):\")\n",
    "print(adata.uns[\"variance_pca_var\"][\"variance_ratio\"])\n",
    "print(\"The explained variance (n_comp):\")\n",
    "print(adata.uns[\"variance_pca_var\"][\"variance\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plot PCA results \n",
    "We can plot the PCA results on a 2D projection, look at the explained var in each PC using the scree plot, and plot the loadings od the PCs, either per PC or a scatter of 2 PCs, to understand their 'drivers'. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axm = create_figure(2, 2, figsize=(12, 12))\n",
    "\n",
    "ax = axm.next()\n",
    "# PCA plot colored by replicate\n",
    "Plots.plot_pca(\n",
    "    data=adata,\n",
    "    ax=ax,\n",
    "    pc_x=1,\n",
    "    pc_y=2,\n",
    "    label=False,\n",
    "    label_column=None,\n",
    "    embbedings_name=None,\n",
    "    color_map_column=\"replicate\",\n",
    ")\n",
    "\n",
    "# scree plot to show the explained variance by each PC\n",
    "ax = axm.next()\n",
    "Plots.scree_plot(adata=adata, ax=ax, n_pcs=50)\n",
    "\n",
    "# top loadings of the first PC\n",
    "ax = axm.next()\n",
    "Plots.plot_pca_loadings(\n",
    "    data=adata,\n",
    "    ax=ax,\n",
    "    dim=1,\n",
    "    nfeatures=20,\n",
    ")\n",
    "\n",
    "# 2d loading plot with highlighted top 20 loadings\n",
    "ax = axm.next()\n",
    "Plots.plot_pca_loadings_2d(\n",
    "    data=adata,\n",
    "    ax=ax,\n",
    "    pc_x=1,\n",
    "    pc_y=2,\n",
    "    nfeatures=20,\n",
    "    add_labels=True,\n",
    "    add_lines=True,\n",
    "    scatter_kwargs=None,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plot PCA results for feature PCA \n",
    "Just like the PCA on the samples, we can plot the same plots for the results of PCA calculated on the features.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# now produce the PCAs plot for the features\n",
    "fig, axm = create_figure(2, 2, figsize=(12, 12))\n",
    "\n",
    "ax = axm.next()\n",
    "Plots.plot_pca(\n",
    "    data=adata,\n",
    "    ax=ax,\n",
    "    pc_x=1,\n",
    "    pc_y=2,\n",
    "    dim_space=\"var\",\n",
    "    label=False,\n",
    "    label_column=None,\n",
    "    embbedings_name=None,\n",
    ")\n",
    "\n",
    "ax = axm.next()\n",
    "Plots.scree_plot(adata=adata, ax=ax, n_pcs=50, dim_space=\"var\")\n",
    "\n",
    "ax = axm.next()\n",
    "Plots.plot_pca_loadings(data=adata, ax=ax, dim=1, nfeatures=10, dim_space=\"var\")\n",
    "\n",
    "ax = axm.next()\n",
    "Plots.plot_pca_loadings_2d(\n",
    "    data=adata,\n",
    "    ax=ax,\n",
    "    pc_x=1,\n",
    "    pc_y=2,\n",
    "    nfeatures=10,\n",
    "    add_labels=True,\n",
    "    add_lines=True,\n",
    "    scatter_kwargs=None,\n",
    "    dim_space=\"var\",\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### UMAP Visualization with Scanpy\n",
    "\n",
    "To explore and visualize the high-dimensional proteomics data, we use **UMAP (Uniform Manifold Approximation and Projection)** as implemented in **Scanpy**. UMAP projects complex, high-dimensional feature spaces into a lower-dimensional space (typically 2D) while preserving the local and global structure of the data. This allows us to identify clusters, relationships, and potential outliers in the proteomic profiles at a glance.\n",
    "\n",
    "In this notebook, Scanpyâ€™s `sc.pp.neighbors()` and `sc.tl.umap()` functions are applied to the processed data matrix to compute a nearest-neighbor graph and then generate UMAP coordinates. We will use the sample PCA matrix for neighbor calculations. The resulting UMAP embedding provides an intuitive visualization of sample similarity and grouping based on proteomic features, complementing downstream analyses such as clustering or differential expression.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "\n",
    "sc.pp.neighbors(adata, n_neighbors=10, use_rep=\"X_pca_obs\")  # use the PCA results on samples\n",
    "sc.tl.umap(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# location of the umap coordinates in the adata.obsm\n",
    "print(\"The UMAP coordinates in the adata.obsm are in adata.obsm['X_umap'] with shape: \", adata.obsm[\"X_umap\"].shape)\n",
    "print(adata.obsm[\"X_umap\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plot UMAP\n",
    "We can either plot the UMAP results using scanpy's plotting function, or we can use AlphaTools plotting function, with adding the umap coordinates directly to the obs df.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# scanpy's plotting function\n",
    "sc.pl.umap(adata, color=\"replicate\", size=50)  # the size is usually much smaller"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Another option is to copy the coordinates into the `adata.obs` data frame, to plot in using `scatter` function in `alphatools` package"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs[\"UMAP1\"] = adata.obsm[\"X_umap\"][:, 0]\n",
    "adata.obs[\"UMAP2\"] = adata.obsm[\"X_umap\"][:, 1]\n",
    "\n",
    "fig, axm = create_figure(1, 1, figsize=(5, 5))\n",
    "ax = axm.next()\n",
    "Plots.scatter(adata, x_column=\"UMAP1\", y_column=\"UMAP2\", color_map_column=\"replicate\", ax=ax, legend=\"auto\")\n",
    "\n",
    "label_axes(\n",
    "    ax=ax,\n",
    "    xlabel=\"UMAP_1\",\n",
    "    ylabel=\"UMAP_2\",\n",
    "    title=\"UMAP colored by replicate (Alphatools implementation)\",\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "alphaverse",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
