{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Example of PELSA differential expression with `alphatools`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## An example dataset: The original PELSA publication\n",
    "\n",
    "PELSA [1] is a novel method to investigate protein-ligand interactions through limited proteolysis. Cell lysate is treated with a short pulse of trypsin at extremely high (1:2) enzyme - substrate ratio, which allows for the digestion of surface exposed peptides. If a ligand (such as a small molecule binder) is bound to the protein surface, it stablizes the surrounding protein region and digestion is momentarily slowed. When compared against a control without the ligand, the PELSA-stabilized peptides appear downregulated. We replicate the original publication's analysis of Staurosporine, a pan-kinase binder, and visualize the regulation of kinase targets.\n",
    "\n",
    "[1]: Li, Kejia, et al. \"A peptide-centric local stability assay enables proteome-scale identification of the protein targets and binding regions of diverse ligands.\" Nature Methods 22.2 (2025): 278-282."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import tempfile\n",
    "\n",
    "import logging\n",
    "from alphatools import io  # input/output, preprocessing, plotting and tools modules\n",
    "from alphabase.tools.data_downloader import DataShareDownloader\n",
    "\n",
    "logger = logging.getLogger(__name__)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Preparing the dataset using `alphatools` loaders and AnnData factory. \n",
    "\n",
    "After downloading the relevant files from the study's PRIDE-repository (https://www.ebi.ac.uk/pride/archive/projects/PXD034606):\n",
    "\n",
    "- LKJ_20211007_480_Hela_stau_0uM_1.raw\n",
    "- LKJ_20211007_480_Hela_stau_0uM_2.raw\n",
    "- LKJ_20211007_480_Hela_stau_0uM_3.raw\n",
    "- LKJ_20211007_480_Hela_stau_0uM_4.raw\n",
    "- LKJ_20211007_480_Hela_stau_20uM_1.raw\n",
    "- LKJ_20211007_480_Hela_stau_20uM_2.raw\n",
    "- LKJ_20211007_480_Hela_stau_20uM_3.raw\n",
    "- LKJ_20211007_480_Hela_stau_20uM_4.raw\n",
    "\n",
    "And processing them with DIANN 2.1.0, the report.parquet file was saved in our datashare alongside sample metadata. We have to extract protein-level intensities and precursor-level intensities from this dataset using the `alphatools.io.AnnDataFactory` class. The resulting AnnData object contains protein-group or precursor quantities and any number of feature-metadata columns (for example, protein groups may have genes as secondary annotation, precursors may have protein groups and genes as secondary annotation)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/var/folders/2l/hhd_z4hx3070zw8rlj4c3l940000gn/T/tmpjf7x973v/report.parquet does not yet exist\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100% |########################################################################|\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/var/folders/2l/hhd_z4hx3070zw8rlj4c3l940000gn/T/tmpjf7x973v/report.parquet successfully downloaded (70.8467435836792 MB)\n"
     ]
    }
   ],
   "source": [
    "report_url = \"https://datashare.biochem.mpg.de/s/6piDQGm2yAEdtKQ/download\"\n",
    "sample_metadata_url = \"https://datashare.biochem.mpg.de/s/NGof744gWw66Mc8/download\"\n",
    "\n",
    "# Get the report by downloading it to a temporary directory and directly loading it from there\n",
    "with tempfile.TemporaryDirectory() as temp_dir:\n",
    "    file_path = DataShareDownloader(url=report_url, output_dir=temp_dir).download()\n",
    "\n",
    "    # AnnDataFactory instance containing protein level data\n",
    "    adata_protein = io.read_psm_table(\n",
    "        file_paths=file_path,\n",
    "        search_engine=\"diann\",\n",
    "        intensity_column=\"PG.MaxLFQ\",\n",
    "        feature_id_column=\"Protein.Group\",\n",
    "        sample_id_column=\"Run\",\n",
    "        secondary_feature_columns=[\"genes\"],\n",
    "    )\n",
    "\n",
    "    # AnnDataFactory instance containing gene level data\n",
    "    adata_gene = io.read_psm_table(\n",
    "        file_paths=file_path,\n",
    "        search_engine=\"diann\",\n",
    "        intensity_column=\"Genes.MaxLFQ\",\n",
    "        feature_id_column=\"Genes\",\n",
    "        sample_id_column=\"Run\",\n",
    "        secondary_feature_columns=[\"proteins\"],\n",
    "    )\n",
    "\n",
    "    # AnnDataFactory instance containing precursor level data\n",
    "    adata_precursor = io.read_psm_table(\n",
    "        file_paths=file_path,\n",
    "        search_engine=\"diann\",\n",
    "        intensity_column=\"Precursor.Normalised\",\n",
    "        feature_id_column=[\"Precursor.Id\", \"Genes\", \"Protein.Group\"],\n",
    "        sample_id_column=\"Run\",\n",
    "        secondary_feature_columns=[\"genes\", \"sequence\"],\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "With our different AnnData objects in hand, we can look at "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "alphaverse",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
