{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Filter and annotate study data based on metadata\n",
    "\n",
    "A particular strength of AlphaTools is its capability for easy filtering and subsetting of data, and assignment of categories based on filtering conditions. _*filter_by_metadata*_ takes easy-to-understand filter dictionaries, while _*assign_classes*_ works on the exact same logic, but instead of filtering the data it assigns classes to samples that meet one or the other filter criteria. Notably, it detects double assignments and thereby prevents overlapping classes.\n",
    "\n",
    "#TODO: integrate PR description"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "--> diagram"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### code ###"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Preprocessing \n",
    "\n",
    "Most publications in the field of proteomics employ a consensus set of preprocessing steps. There are statistical and scientific motivations for each of them, and AlphaTools offers tested and validated implementations of each of them. For some, like PCA, we closely align with ScanPy's PCA functions. For others, like imputation, we provide an implementation that fits the different underlying distribution of proteomics vs. transcriptomics data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### code ###\n",
    "\n",
    "# log2 transform\n",
    "# missing value imputation\n",
    "# scaling/centering\n",
    "# pca"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5. Visualization\n",
    "\n",
    "AlphaTools offers a lightweight interface to matplotlib to generate publication-quality figures. Default parameters and plotting functions are chosen to be intuitive and straightforward, however, the code remains fully open and customizable. A default style and color palette is set, with different options for categorical and continuous coloring. (see tutorial notebook t02_plotting.ipynb for further reference). Saving plots and the data to generate the plots is straightforward, ensuring reproducible visualizations.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### code ###\n",
    "\n",
    "# scatterplot of pca components, barplot of explained variances, lineplot of loadings\n",
    "\n",
    "#"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 6. Differential expression\n",
    "\n",
    "An array of hypothesis tests is offered, with the default being a two-sided two-sample t-test that does not assume equal variance (MWU). P-values are adjusted using BH-FDR, and optionally permutation based significance cutoffs calculated (paper ref)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### code ###"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 7. Visualize fold-changes, p-values and cutoffs\n",
    "\n",
    "A strength of AlphaTools is its modular design, as exemplified by the volcanoplot below: rather than providing a big black-box function that takes data and spits out a plot, the individual elements are added by the user. This means that the visualized dataset is always exposed and can be saved & manipulated easily with custom annotations and other plot elements"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### code ### volcanoplots"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
